Feat: Axis-based Image Stretching

arul-trenser opened the pull request:
<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context

- PR for the Feature Request: **Axis-based Image Stretching** #2287 
- This PR has been incorporated by [FlyWheel.io](https://flywheel.io/).

<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->

### Changes & Results
This PR introduces **axis-based image stretching** using camera projection matrix manipulation.

#### Usage Example
**Stretch along the X-axis:**
```js
const { aspectRatio }  = viewport.getCamera()
viewport.setCamera({
    acpectRatio: [2, aspectRatio[1]]
})
```

**Stretch along the X-axis:**
```js
const { aspectRatio }  = viewport.getCamera()
viewport.setCamera({
    acpectRatio: [aspectRatio[0], 2]
})
```

<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?
- Before vs After
- Screenshots / GIFs / Videos
-->

### Testing

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->

Basic Stretching along X and Y axis:

https://github.com/user-attachments/assets/20b3d859-fd03-4a7e-ac22-c08bd811723f

Segmentation Brush Tool Behavior in Stretched Viewport:

https://github.com/user-attachments/assets/9a5162e7-5ebe-496d-8f74-bbdd80ece85a

Circle ROI Tool Behavior in Stretched Viewport

https://github.com/user-attachments/assets/17ced2a7-09a9-47f3-879b-f3c80621c331

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://cornerstonejs.org/ -->

- [] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] "OS: <!--[e.g. Windows 10, macOS 10.15.4]"--> Windows 11
- [x] "Node version: <!--[e.g. 16.14.0]"--> 22.19.0
- [x] "Browser: Chrome  141.0.7390.123
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]"-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

arul-trenser committed:
feature for axis based stretching

arul-trenser committed:
feature for axis based stretching

arul-trenser committed:
fix unit tests

arul-trenser committed:
fix unit tests

arul-trenser committed:
Normalize pan to remove effect of stretch

arul-trenser committed:
Normalize pan to remove effect of stretch

sedghi commented:
how we can test it?

sedghi commented on packages/core/src/types/CPUFallbackViewport.ts:
```diff
@@ -26,6 +27,7 @@ interface CPUFallbackViewport {
   colormap?: CPUFallbackColormap;
   displayedArea?: CPUFallbackViewportDisplayedArea;
   modality?: string;
+  aspectRatio?: Point2;
```
are you adding aspect ratio for both cpu and gpu pipelines?

arul-trenser committed:
fix: radii calc

arul-trenser committed:
fix: radii calc

arul-trenser committed:
removed aspectRatio property in CPUFallbackViewport

arul-trenser committed:
removed aspectRatio property in CPUFallbackViewport

arul-trenser committed:
Added example for axialBasedImageStretching

arul-trenser committed:
Added example for axialBasedImageStretching

arul-trenser committed:
fixed build issue

arul-trenser committed:
fixed build issue

arul-trenser commented:
> how we can test it?

@sedghi  Added an example demonstrating the stretch behavior in `axialBasedImageStretching`
Also attached videos in the PR description

jmannau commented:
Hi,
We work with systems that have different vertical/horizontal scales.  For example, with Optical Coherence Topography (https://eyewiki.org/Optical_Coherence_Tomography) scans, it is common to display the image at 2:1 (vertical/horizontal) aspect ration.  The scans often have 2x->4x vertical resolution compared to the horizontal scale. This allows better visualisation of pathology in the vertical/depth axis. We also allow users to change the aspect ratio between 1:1 and 4:1 to account for different vendor devices and scan types. 
Below are 2 images that demonstrate this.

**1:1 Aspect Ratio**
<img width="688" height="231" alt="Screenshot 2025-11-12 at 11 23 17 am" src="https://github.com/user-attachments/assets/eb5281f3-b53e-4f0b-9ee5-6b7c18fedb6d" />

**the same scan at 2:1 Aspect Ratio**
<img width="684" height="457" alt="Screenshot 2025-11-12 at 11 23 30 am" src="https://github.com/user-attachments/assets/15ed0ddb-9c90-4986-92c3-46726b00c5e2" />

jmannau commented:
Happy to. But we’re very short on time until the new year. I’ll see what we can do

wayfarer3130 commented on packages/core/src/RenderingEngine/StackViewport.ts:
```diff
@@ -2683,7 +2691,7 @@ class StackViewport extends Viewport {
   }
 
-  private resetCameraGPU({ resetPan, resetZoom }): boolean {
+  private resetCameraGPU({ resetPan, resetZoom, resetAspectRatio }): boolean {
```
Can you just pass the reset options to the parent rather than needing to specifically list them?  It means that all new reset options will just flow through unless specifically editted.
eg:
resetCameraGPU(resetOptions?: ResetOptions): boolean

...
return super.resetCamera({
   resetToCenter,
   ...resetOptions
})

wayfarer3130 commented on packages/core/src/RenderingEngine/Viewport.ts:
```diff
@@ -1391,6 +1396,36 @@ class Viewport {
+   *          based on the active camera settings.
+   */
+  public getAspectRatio(): Point2 {
+    const { aspectRatio } = this.getCamera();
```
@sedghi - what are your thoughts on the initial aspect ratio being a property of the viewport, and then the current aspect ratio gets applied later?  This interacts with some of our thoughts on how to manage the viewport state in OHIF.

arul-trenser replied:
Added aspectRatio property for viewport.

wayfarer3130 commented on packages/tools/src/tools/annotation/CircleROITool.ts:
```diff
@@ -238,14 +243,84 @@ class CircleROITool extends AnnotationTool {
-    const radiusPoint = getCanvasCircleRadius([canvasCenter, canvasCoords]);
+    // Get the radius in world units from the drag distance
+    const ellipseWorldCoordinates = getEllipseWorldCoordinates(
+      [points[0], points[1]],
```
Just pass in points or points.slice(0,1), but I believe you will only have two points.

wayfarer3130 commented:
We need to define some requirements as part of this change for what it is that the circle, sphere and elliptical tools are doing.
The sphere I believe should always be a sphere in world coordinates, as otherwise it does weird things to the shape.  That means it should show as an ellipse if displaying world coordinates does not result in an aspect ratio of 1:1 for the rows/columns.

Circle and circle brushes are harder because of the differences in how things get saved.  The circle brush is never saved in image coordinates, so I think we could fairly easily display a circle in the current transformed coordinates.
Circle annotations are saved as both MPR in world coordinates and as index coordinates.
This needs some documentation in a requirements section for exactly what it is that is meant by "sphere" "circle brush" and "circle tool" under MPR/co-planar modes with index, canvas and world coordinates at different aspect ratios.

arul-trenser committed:
Review comments updated

arul-trenser committed:
Review comments updated

arul-trenser committed:
Added aspectRatio property to viewport

arul-trenser committed:
Added aspectRatio property to viewport

arul-trenser committed:
updated getCenterAndRadiusInCanvas to use world and canvas point

arul-trenser committed:
updated getCenterAndRadiusInCanvas to use world and canvas point

arul-trenser committed:
Added documentation for tools

arul-trenser committed:
Added documentation for tools

arul-trenser commented:
@wayfarer3130 Thanks for the feedback! I’ve applied the suggested changes and included the documentation. Looking forward to your thoughts; please review and approve if everything looks good.

wayfarer3130 committed:
Merge remote-tracking branch 'origin/main' into feat-axis-based-image-stretching

wayfarer3130 committed:
Add aspect ratio testing for resize and calibration

wayfarer3130 committed:
Add aspect ratio testing for resize and calibration

wayfarer3130 commented on packages/core/src/RenderingEngine/Viewport.ts:
```diff
@@ -1454,7 +1513,7 @@ class Viewport {
 
   protected getCameraNoRotation(): ICamera {
-    const vtkCamera = this.getVtkActiveCamera();
+    const vtkCamera = this.getVtkActiveCamera() as extendedVtkCamera;
```
@sedghi - how can the default return type of getVtkActiveCamera be setup so that it has the extended methods for aspect ratio?

wayfarer3130 commented on packages/tools/examples/resize/index.ts:
This allows both the viewport size and the canvas aspect ratio to be varied independently and then resizing applied.  It demonstrates that the size changes aren't working correctly any longer with resize.
This can be seen by switching between the 0.5:1 and 1:2 ratios - both of these should appear exactly identical as they only affect aspect ratio, not zoom, but right now they end up modifying the zoom.

arul-trenser replied:
Fixed the issue.

https://github.com/user-attachments/assets/3a81ff51-1e07-4836-866a-4d464a2283ed

wayfarer3130 commented:
@sedghi - I think now that it is clarified that the aspect ratio applies only to the display and not to the coordinate/world space, I think this becomes a lot like rotation or flip.  There are still some bugs to fix in terms of the resize operation relative to the aspect ratio, where it no long applies the zoom to fit or scale size.

arul-trenser commented on packages/core/src/RenderingEngine/Viewport.ts:
```diff
@@ -1391,6 +1398,58 @@ class Viewport {
+   *          based on the active camera settings.
+   */
+  public getAspectRatio(): Point2 {
+    if (this.options?.aspectRatio) {
```
Updated to return the correct aspectRatio

arul-trenser committed:
fix review comments

arul-trenser committed:
fixed the resize issue for aspectRatio

arul-trenser committed:
Modified the getProjectionScaleIndices to getProjectScaleMatrix

arul-trenser committed:
Fixed unit test

arul-trenser committed:
fixed review comments

sen-trenser commented:
@wayfarer3130 Could you please review and share your feedback on this changes?

arul-trenser commented:
@wayfarer3130 Could you please review and share your feedback for this PR?

wayfarer3130 committed:
Rename axial to axis

wayfarer3130 committed:
Add rotation button

wayfarer3130 committed:
Add stack viewport to test on stack views.

wayfarer3130 commented on packages/core/src/RenderingEngine/Viewport.ts:
None of the aspect ratio stuff applies to the CPU based rendering - there isn't any reason it can't, and it is important to allow screen captures/display with such information on it.

wayfarer3130 reviewed (changes requested):
The rotation of images is such a fundamental part of MPR viewing that I think it needs to be fixed.
Also, the scale of the image area should be updated so that scale to fit applies the specified transform on the image aspect so as to scale to fit after the transform.  If you can figure out scale to fit fixes for rotated images at the same time, I wouldn't mind having that included as well, but it isn't directly part of this PR so it doesn't need to be included.

arul-trenser committed:
Fixed the rotation issue, used function isEqual and  added stretch for CPU rendering

arul-trenser committed:
fixed unit testing

arul-trenser committed:
canvas stretching

arul-trenser committed:
fix unit test

arul-trenser committed:
modified comment

arul-trenser committed:
updated docs

arul-trenser committed:
Merge branch 'cornerstonejs:main' into feat-axis-based-image-stretching

arul-trenser commented:
@wayfarer3130
I have addressed all the review comments.

The implementation has been updated to use canvas-space stretching, ensuring that the stretch is applied relative to the canvas axes regardless of the MPR view rotation. This maintains consistent stretching behavior even when the image is rotated.

Could you please re-review the PR and share your feedback?

wayfarer3130 committed:
Merge branch 'feat-axis-based-image-stretching' of https://github.com/arul-trenser/cornerstone3D into feat-axis-based-image-stretching

wayfarer3130 committed:
fix: Deleted unused code.  Added testing capability to axis based image stretching

wayfarer3130 reviewed (changes requested):
It is looking close - I added the ability to test with CPU rendering, and to have different viewport types, with the crosshairs bound to ctrl+click to allow orientation changes.

I am still noticing the changing the aspect ratio does not fit to viewport.  I suspect you need a reset camera operation that has a don't change aspect ratio and that it should get called when you change the aspect ratio.  Then, the fit to viewport needs to take into account the updated aspect ratio size.

wayfarer3130 committed:
Merge remote-tracking branch 'origin/main' into feat-axis-based-image-stretching

sen-trenser commented:
@wayfarer3130  
> I am still noticing the changing the aspect ratio does not fit to viewport.

> Then, the fit to viewport needs to take into account the updated aspect ratio size.

Keeping the image fitted to the viewport after applying aspect ratio changes would cause the image to shrink rather than stretch. The intended behavior of the stretch feature is to keep the image centered and expand it outward along the selected axis while preserving the original size of the non-stretched axis. This allows users to view and annotate details more accurately in parallel to the stretched direction.

Could you share the use case for refitting the image to the viewport after stretching? Understanding this use case will help us assess whether supporting this behavior as the default would be beneficial to end users.

Thanks!

cc: @kgoettler

arul-trenser committed:
Fix: review comment by adding a config for fix to viewport after stretch

arul-trenser commented:
> changing the aspect ratio does not fit to viewport.

Fixed the above scenario  by adding a configuration option to control whether the viewport should fit after applying the stretch. The current default behavior is to fit the viewport after stretch.

@wayfarer3130 Could you please take a look at this PR and provide your feedbacks?

wayfarer3130 commented:
@arul-trenser - could you fix the OHIF downstream tests?  This should be passing for those.

wayfarer3130 commented:
> @wayfarer3130
> 
> > I am still noticing the changing the aspect ratio does not fit to viewport.
> 
> > Then, the fit to viewport needs to take into account the updated aspect ratio size.
> 
> Keeping the image fitted to the viewport after applying aspect ratio changes would cause the image to shrink rather than stretch. The intended behavior of the stretch feature is to keep the image centered and expand it outward along the selected axis while preserving the original size of the non-stretched axis. This allows users to view and annotate details more accurately in parallel to the stretched direction.
> 
> Could you share the use case for refitting the image to the viewport after stretching? Understanding this use case will help us assess whether supporting this behavior as the default would be beneficial to end users.
> 
> Thanks!
> 
> cc: @kgoettler

The point is to be able to visually see fractures or other features that are narrow in the direction of stretch and wide orthogonal  For example, rib fractures.   If the zoom factor is reset so that the image stays in view, then the user can then use the zoom and pan tools to go through the entire image to look for such features.    If they have adjusted it and don't initially see the full image, it is possible that a small portion of the image will not be visible AND will not be noticed to be offscreen.  So, in terms of safety I would prefer to start with displaying the full image after changing the aspect ratio.
If one really wants a single-click tool, then probably the best option would be a way to draw a line that automatically orients itself up-down on the screen or left/right depending on the original orientation, AND stretches the line to 2:1 or something like that.  That way the clinician would know exactly what they are seeing, and if they wanted a full-image view, then the line would start out full-range.  For now, I'd like the scale to fit to work to prevent issues.

arul-trenser committed:
fix: OHIF downstream test related to brush radius

arul-trenser committed:
Merge branch 'cornerstonejs:main' into feat-axis-based-image-stretching

wayfarer3130 committed:
Merge remote-tracking branch 'origin/main' into feat-axis-based-image-stretching

wayfarer3130 reviewed (approved):
Thank you for sticking with this - it has taken a while to get this through. 
Will just waiting for all the tests to complete - I updated to origin/main

arul-trenser committed:
fix: issue with stretch and scroll leads to zoom in or zoom out the image

arul-trenser commented:
@wayfarer3130  Hi, I faced an issue where the scroll wasn't working properly with the 'stretch + fit to viewport' setting. I have fixed the scrolling issue in commit [7ad379c](https://github.com/cornerstonejs/cornerstone3D/pull/2443/commits/7ad379c518286be0270091ac543224fa355c8e7b). Could you please review the changes? Apologies for missing this earlier. 

Thanks in advance!

arul-trenser committed:
Merge branch 'cornerstonejs:main' into feat-axis-based-image-stretching
